### PR TITLE
Restructuring 2 - Add Utility functions and Run all Tests

### PR DESCRIPTION
**Pull request by kartikdutt18**
This PR is transferred from #8, Since FFN weights some time cause invalid memory access which seems related to mlpack/mlpack#2314. So I am adding utility function to unzip files here and I'll continue working on image dataloader and then we can get back to the LeNet model.
I hope this makes sense.
This PR provides the following :
1. Unzip Utility Function.
2. Some bug fixes.
3. Runs all tests. 
(Probably we should get mnist_test/train in mnist.tar.gz since mnist test and train are fixed).
4. ListDir Utility function.

Looking forward to your suggestions.
Regards.

CC: @KimSangYeon-DGU

### CODE REVIEW

**Comment by kartikdutt18** [comment]
This works now on windows too now, Cleaning commits and outputs.

**Comment by zoq** [comment]
>  (Probably we should get mnist_test/train in mnist.tar.gz since mnist test and train are fixed).

Sure, we can do that, is there anything else you like to add here, this looks pretty good to me.

**Comment by kartikdutt18** [comment]
> Sure, we can do that, is there anything else you like to add here, this looks pretty good to me.

I am done with this PR from my side aside from placing mnist_train/test in mnist.tar.gz. If this makes sense I'll ask on the IRC once I incorporate all suggestions.

**Comment by zoq** [comment]
Makes sense to me.

**Comment by KimSangYeon-DGU** [comment]
Looks good to me as well, please let us know if this is ready to go :)

File: tests/dataloader_tests.cpp [hunk 1]
```diff
@@ -50,19 +50,23 @@ BOOST_AUTO_TEST_CASE(CSVDataLoaderTest)
...
 BOOST_AUTO_TEST_CASE(MNISTDataLoaderTest)
 {
+  /**
   DataLoader<> dataloader("mnist", true, 0.80);
   // Check for correct dimensions.
   BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_cols, 784);
   BOOST_REQUIRE_EQUAL(dataloader.TestFeatures().n_cols, 784);
   BOOST_REQUIRE_EQUAL(dataloader.ValidFeatures().n_cols, 784);
   BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_rows, 33600);
+  */
```
Comment (KimSangYeon-DGU) [inline tests/dataloader_tests.cpp:69]: Should we keep this commented block?

File: utils/utils.hpp [hunk 2]
```diff
@@ -213,5 +264,35 @@ class Utils
...
+      std::copy(boost::filesystem::directory_iterator(directoryPath),
+          boost::filesystem::directory_iterator(),
+          std::back_inserter(pathVector));
+
+      // Sort the path vector.
+      std::sort(pathVector.begin(), pathVector.end());
+    }
+    else
+    {
+      mlpack::Log::Warn << "The " << path << "Doesn't exist." << std::endl;
```
Comment (KimSangYeon-DGU) [inline utils/utils.hpp:294]:
```suggestion
      mlpack::Log::Warn << "The " << path << " doesn't exist." << std::endl;
```

File: tests/dataloader_tests.cpp (hunk 1 above)
Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: Once the mnist_train / test.csv in in mlpack.org/datasets we can uncomment this.

File: utils/utils.hpp (hunk 2 above)
Comment (kartikdutt18) [reply utils/utils.hpp:294]: Right, we fix it in the next commit after I rebase. We'll have a green build. Thanks a lot.

**Comment by kartikdutt18** [comment]
This is ready from my side. Just tested it everything works fine. I think the builds will result in the same output.
Regards.

**Comment by kartikdutt18** [comment]
Sorry pushed in the wrong branch, reverting now.

File: tests/dataloader_tests.cpp [hunk 3]
```diff
@@ -58,11 +60,28 @@ BOOST_AUTO_TEST_CASE(CSVDataLoaderTest)
...
+
+
+  BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_cols, 8400);
+  BOOST_REQUIRE_EQUAL(dataloader.ValidFeatures().n_cols, 33600);
+  BOOST_REQUIRE_EQUAL(dataloader.TestFeatures().n_cols, 28000);
+
+  BOOST_REQUIRE_EQUAL(std::get<0>(dataloader.TrainSet()).n_cols, 8400);
+  BOOST_REQUIRE_EQUAL(std::get<1>(dataloader.TrainSet()).n_rows, 1);
+  BOOST_REQUIRE_EQUAL(std::get<0>(dataloader.ValidSet()).n_cols, 33600);
+  BOOST_REQUIRE_EQUAL(std::get<1>(dataloader.TrainSet()).n_rows, 1);
```
Comment (KimSangYeon-DGU) [inline tests/dataloader_tests.cpp:77]: We also need to add a comment for this block

File: tests/dataloader_tests.cpp [hunk 4]
```diff
@@ -58,11 +60,28 @@ BOOST_AUTO_TEST_CASE(CSVDataLoaderTest)
...
   // Check for correct dimensions.
-  BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_cols, 784);
-  BOOST_REQUIRE_EQUAL(dataloader.TestFeatures().n_cols, 784);
-  BOOST_REQUIRE_EQUAL(dataloader.ValidFeatures().n_cols, 784);
-  BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_rows, 33600);
+  BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_rows, 784);
+  BOOST_REQUIRE_EQUAL(dataloader.TestFeatures().n_rows, 784);
+  BOOST_REQUIRE_EQUAL(dataloader.ValidFeatures().n_rows, 784);
+
+
```
Comment (KimSangYeon-DGU) [inline tests/dataloader_tests.cpp:69]: Can you use 1 newline and add explanation about what we're checking in this block?

And, perhaps, I missed something... can you let me know why the data were distributed like below?
`Train : 8,400`
`Valid: 33,600`
`Test: 28,000`

File: utils/utils.hpp [hunk 5]
```diff
@@ -213,5 +264,35 @@ class Utils
...
+   * Fills a vector with paths to all files in directory.
+   *
+   * @param path Path to Directory.
+   * @param pathVector A vector of type filesystem::path, which will be filled
+   *                   paths for all files / folders in given directory path.
+   * @param absolutePath Boolean to determine if path is absolute or relative.
+   */
+  static void ListDir(const std::string& path,
+                      std::vector<boost::filesystem::path>& pathVector,
+                      const bool absolutePath = false)
```
Comment (KimSangYeon-DGU) [inline utils/utils.hpp:278]: Can you let me know the use case of this function?

File: tests/dataloader_tests.cpp (hunk 3 above)
Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:77]: Will do in the next commit. Thanks.

File: utils/utils.hpp (hunk 5 above)
Comment (kartikdutt18) [reply utils/utils.hpp:278]: This fills the pathVector with paths of all files in the `path`. For example, if we have a folder called images with 30 files with names 1.jpeg, 2.jpeg and so on then pathVector will be equal to `{"1.jpeg", "2.jpeg" ...}`. I use it to get all annotations and images in the folder for PASCAL VOC dataloader.

Comment (KimSangYeon-DGU) [reply utils/utils.hpp:278]: Aha, thanks for letting me know! nice :)

File: tests/dataloader_tests.cpp (hunk 4 above)
Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: This is the same split as mlpack::data::split. Here 0.8 means 80% testing and 20 % training data. mlpack takes in test ratio. I though it would be best not to invert. Let me know if I should do that or keep it same as mlpack.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp:69]: Ok, then, it would be better to change the parameter name from `ratio` to `testRatio` for clarification :)

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: Sure, I'll make the change.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp:69]: Also, kindly update the parameter description as well in [here](https://github.com/kartikdutt18/models/blob/6ae432e2951f2a222fc37fe796b7131dea0e9da1/dataloader/dataloader.hpp#L88)

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: Sure, will make the changes. Thanks a lot.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp:69]: It seems [this](https://github.com/kartikdutt18/models/blob/6ae432e2951f2a222fc37fe796b7131dea0e9da1/dataloader/dataloader.hpp#L109) also needs to be changed

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: I think I can set it to 0.25 instead. I think that makes sense.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp:69]: Yes, but I think the `ratio` is a little vague, so it should be changed to `testRatio` and description is also needed to be clear that is a ratio for test data.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp:69]: Ah, but it's not the work for this PR... sorry. I can do it.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp:69]: I can make the change. I almost done with this commit.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: Done, the tests pass locally. The build should reciprocate the same. Opening the other PR now.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: I probably missed something, the mlpack::data::split already uses testRatio and I changed it here too. Kindly let me know if I missed something.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp]: I requested the changes because I think the parameter name `ratio` in `DataLoader` is passed to `mlpack::data::Split()` that uses `testRatio` and the name `ratio` in `DataLoader` is a little vague because I seems train-test split ratio, not test-train ratio. Please let me know if it doesn't make sense.

Also, I'm still trying to find why the data distribution ratio is.
```
Train : 8,400 (12%)
Valid: 33,600 (48%)
Test: 28,000 (40%)
```

The total number of data is `70,000` and if we split the whole data with the test ratio `0.8`, the train-test ratio should be `0.2` and `0.8`. And the current `DataLoader` implementation splits the train data into train and valid data using the test ratio.

So, shouldn't the data ratio be like this? Are the sources of train and test data different?
``
Train: 4%
Valid: 16%
Test: 80%
``

File: tests/dataloader_tests.cpp [hunk 6]
```diff
@@ -58,11 +60,30 @@ BOOST_AUTO_TEST_CASE(CSVDataLoaderTest)
...
+  BOOST_REQUIRE_EQUAL(dataloader.TrainFeatures().n_cols, 8400);
+  BOOST_REQUIRE_EQUAL(dataloader.ValidFeatures().n_cols, 33600);
+  BOOST_REQUIRE_EQUAL(dataloader.TestFeatures().n_cols, 28000);
+
+  // Check if we can access both features and labels using
+  // TrainSet tuple and ValidSet tuple.
+  BOOST_REQUIRE_EQUAL(std::get<0>(dataloader.TrainSet()).n_cols, 8400);
+  BOOST_REQUIRE_EQUAL(std::get<1>(dataloader.TrainSet()).n_rows, 1);
+  BOOST_REQUIRE_EQUAL(std::get<0>(dataloader.ValidSet()).n_cols, 33600);
+  BOOST_REQUIRE_EQUAL(std::get<1>(dataloader.TrainSet()).n_rows, 1);
```
Comment (KimSangYeon-DGU) [inline tests/dataloader_tests.cpp:79]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<1>(dataloader.ValidSet()).n_rows, 1);
```

File: tests/dataloader_tests.cpp (hunk 4 above)
Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: > So, shouldn't the data ratio be like this? Are the sources of train and test data different?

Yes, the sources are different. The train and validation set from mnist_train.csv which contains 40k data points and test comes from mnist_test.csv which has 28k points.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: Often people use test and validation set interchangeably however they are not the same. Validation set comes from train test split where as test data is data on which we actually use the model we trained. Kindly let me know if this makes sense.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp]: Ah, ok. I thought it from the same file :)

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: Should I use validRatio instead I think that makes even more sense.

Comment (KimSangYeon-DGU) [reply tests/dataloader_tests.cpp]: Nice. that makes sense to me as well.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: Right, making the changed.

Comment (kartikdutt18) [reply tests/dataloader_tests.cpp]: Done. :)

**Comment by kartikdutt18** [comment]
Great. Thanks a lot for the reviews.

**Comment by KimSangYeon-DGU** [comment]
Thanks! @kartikdutt18 👍

**Comment by kartikdutt18** [comment]
Thanks @KimSangYeon-DGU, @zoq for the reviews and all the help.
